game-app/game-headless/src/main/java/org/triplea/game/server/HeadlessServerSetup.java

```diff
@@ -5,22 +5,32 @@
 import games.strategy.engine.framework.startup.mc.ServerModel;
 import games.strategy.engine.framework.startup.ui.panels.main.game.selector.GameSelectorModel;
 import java.util.Optional;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
 import org.triplea.game.chat.ChatModel;
 import org.triplea.game.startup.SetupModel;
+import org.triplea.java.Interruptibles;
 
 /** Server setup model. */
 class HeadlessServerSetup implements IRemoteModelListener, SetupModel {
   private final ServerModel model;
   private final GameSelectorModel gameSelectorModel;
+  private final ReentrantLock lock;
+  private final Condition playersUpdated;
+
+  private boolean isRunning = true;
 
   HeadlessServerSetup(final ServerModel model, final GameSelectorModel gameSelectorModel) {
     this.model = model;
     this.gameSelectorModel = gameSelectorModel;
     this.model.setRemoteModelListener(this);
+    lock = new ReentrantLock();
+    playersUpdated = lock.newCondition();
   }
 
   @Override
   public void cancel() {
+    signalPotentialStart(false);
```

Comment:
Is it normal that after cancel() is called, a playerListChanged() will set isRunning back to true?
That seems a bit unintuitive to me, but I guess it depends what `isRunning` is supposed to mean.

Should it instead be `cancelled` (with inverted value)? `isRunning` implies to me that the `waitUntilStart()` loop is active, but that doesn't seem like it would be the case if `cancel` breaks us out of that loop and then the player list changes?

Other than that, logic looks good to me.

Reply:
Good point. I mainly wanted to re-use code but I can see that after cancel we probably don't want the code to un-cancel itself

Reply:
But yeah, the only reason this variable exists is that any waiting threads are properly woken up when the system is shut down.